### PR TITLE
Provide helpful error-message for BindException on quarkus.http.port

### LINKED ISSUE #4890: Improving the BindAddress error
i often see in labs people hitting a BindAddress problem. They tend to blame Quarkus but it is invariably :

- another process running in the same port (usually the http port)
- sometimes a different Quarkus instance running 

We could improve the situation by : 
1. being more explicit in the exception (handling) and proposing ways to discover the culprit process (commands to run)
2. Try and find the process ourselves depending on the OS: something like Process 1234 - myFavAV already uses this port

### PR DESCRIPTION
**Pull request by xumk**
Fixes #4890

### CODE REVIEW

**Comment by xumk** [comment]
@geoand @gastaldi review please

**Review by gastaldi — DISMISSED** [review]
LGTM

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java [hunk 1]
```diff
@@ -108,9 +109,28 @@ public void accept(Integer integer) {
...
+                    log.error("Port " + port + " seems to be in use by another process. " +
+                            "Quarkus may already be running or the port is used by another application.");
+                    if (System.getProperty("os.name").startsWith("Windows")) {
+                        log.info("Use 'netstat -a -b -n -o' to identify the process occupying the port.");
+                        log.info("You can try to kill it with 'taskkill /PID <pid>' or via the Task Manager.");
+                    } else {
+                        log.info("Use 'netstat -anop | grep " + port + "' to identify the process occupying the port.");
+                        log.info("You can try to kill it with 'kill -9 <pid>'.");
+                    }
+                    System.exit(-1);
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:128]: What's the deal with this?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java [hunk 2]
```diff
@@ -108,9 +109,28 @@ public void accept(Integer integer) {
...
                 firstStartCompleted = true;
             } catch (Throwable t) {
-                deploymentProblem = t;
-                if (!augmentDone) {
-                    log.error("Failed to start quarkus", t);
+                Throwable rootCause = t;
+                while (rootCause.getCause() != null) {
+                    rootCause = rootCause.getCause();
+                }
+                if (rootCause instanceof BindException) {
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:116]: These error messages are super useful, thanks!

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 1 above)
Comment (gsmet) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:128]: And shouldn't we call `close();`?

Comment (gastaldi) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:128]: Good catch. I think `deploymentProblem = rootCause;` needs to be assigned here and `System.exit` shouldn't be explicitly called?

**Comment by gastaldi** [comment]
The commits need to be squashed into a single one here

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java]: Yeah, I don't see why `System.exit` is being called here. Can you explain?

Comment (xumk) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java]: @gastaldi ok. I added it

**Comment by xumk** [comment]
@geoand @gastaldi I fix this

**Review by geoand — APPROVED** [review]
Thanks!

**Comment by gastaldi** [comment]
This only works with `mvn quarkus:dev`. I was expecting it to work with `java -jar` and this is the output:

```
/tmp/greeting-quarkus ᐅ java -jar target/greeting-quarkus-0.1.0-runner.jar
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-10-13 13:41:18,229 ERROR [io.qua.run.Application] (main) Failed to start application (with profile prod): java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:609)
	at io.quarkus.runtime.Application.start(Application.java:90)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:93)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:62)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:104)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
Caused by: java.lang.RuntimeException: Unable to start HTTP server
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:501)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.startServer(VertxHttpRecorder.java:224)
	at io.quarkus.deployment.steps.VertxHttpProcessor$openSocket-2064782366.deploy_0(VertxHttpProcessor$openSocket-2064782366.zig:136)
	at io.quarkus.deployment.steps.VertxHttpProcessor$openSocket-2064782366.deploy(VertxHttpProcessor$openSocket-2064782366.zig:40)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:543)
	... 6 more
Caused by: java.util.concurrent.ExecutionException: java.net.BindException: Address already in use
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:467)
	... 10 more
Caused by: java.net.BindException: Address already in use
	at java.base/sun.nio.ch.Net.bind0(Native Method)
	at java.base/sun.nio.ch.Net.bind(Net.java:455)
	at java.base/sun.nio.ch.Net.bind(Net.java:447)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:227)
	at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:134)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:550)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
	at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
	at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973)
	at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:248)
	at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasksFrom(SingleThreadEventExecutor.java:428)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:377)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:488)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by geoand** [comment]
Dev mode is where would normally encounter this error anyway

**Comment by xumk** [comment]
@gastaldi I think this task #4890 only dev mode. Add this messages for `java -jar` its PR or create new PR?

**Comment by gastaldi** [comment]
It seems that still needs some adjustments. Here is the output of mvn quarkus:dev when another process is running in 8080 already: 

```shell

/tmp/greeting-quarkus ᐅ mvn quarkus:dev                                    
[INFO] Scanning for projects...
[INFO] 
[INFO] --------------------< micro.apps:greeting-quarkus >---------------------
[INFO] Building greeting-quarkus 0.1.0
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:dev (default-cli) @ greeting-quarkus ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /tmp/greeting-quarkus/target/classes
Listening for transport dt_socket at address: 5005
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-10-14 13:35:22,316 ERROR [io.qua.run.Application] (Quarkus Main Thread) Port 8080 seems to be in use by another process. Quarkus may already be running or the port is used by another application.
2020-10-14 13:35:22,317 INFO  [io.qua.run.Application] (Quarkus Main Thread) Use 'netstat -anop | grep 8080' to identify the process occupying the port.
2020-10-14 13:35:22,318 INFO  [io.qua.run.Application] (Quarkus Main Thread) You can try to kill it with 'kill -9 <pid>'.
Quarkus application exited with code 1
Press Enter to restart or Ctrl + C to quit

Restarting...
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-10-14 13:35:27,425 ERROR [io.qua.run.Application] (Quarkus Main Thread) Port 8080 seems to be in use by another process. Quarkus may already be running or the port is used by another application.
2020-10-14 13:35:27,426 INFO  [io.qua.run.Application] (Quarkus Main Thread) Use 'netstat -anop | grep 8080' to identify the process occupying the port.
2020-10-14 13:35:27,426 INFO  [io.qua.run.Application] (Quarkus Main Thread) You can try to kill it with 'kill -9 <pid>'.
2020-10-14 13:35:27,425 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (Quarkus Main Thread) Failed to start quarkus: java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.dev.appstate.ApplicationStateNotification.waitForApplicationStart(ApplicationStateNotification.java:51)
	at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:145)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartApp(IsolatedDevModeMain.java:181)
	at io.quarkus.deployment.dev.IsolatedDevModeMain$1.accept(IsolatedDevModeMain.java:99)
	at io.quarkus.deployment.dev.IsolatedDevModeMain$1.accept(IsolatedDevModeMain.java:81)
	at io.quarkus.runtime.ApplicationLifecycleManager.lambda$setDefaultExitCodeHandler$0(ApplicationLifecycleManager.java:271)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:175)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:62)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:104)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.runner.bootstrap.StartupActionImpl$3.run(StartupActionImpl.java:134)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:675)
	at io.quarkus.runtime.Application.start(Application.java:90)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:95)
	... 10 more
Caused by: java.lang.RuntimeException: Unable to start HTTP server
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:501)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.startServer(VertxHttpRecorder.java:224)
	at io.quarkus.deployment.steps.VertxHttpProcessor$openSocket-2064782366.deploy_0(VertxHttpProcessor$openSocket-2064782366.zig:136)
	at io.quarkus.deployment.steps.VertxHttpProcessor$openSocket-2064782366.deploy(VertxHttpProcessor$openSocket-2064782366.zig:40)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:620)
	... 12 more
Caused by: java.util.concurrent.ExecutionException: java.net.BindException: Address already in use
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:467)
	... 16 more
Caused by: java.net.BindException: Address already in use
	at java.base/sun.nio.ch.Net.bind0(Native Method)
	at java.base/sun.nio.ch.Net.bind(Net.java:455)
	at java.base/sun.nio.ch.Net.bind(Net.java:447)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:227)
	at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:134)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:550)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
	at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
	at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973)
	at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:248)
	at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasksFrom(SingleThreadEventExecutor.java:428)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:377)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:488)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more
```

**Comment by xumk** [comment]
@gastaldi ah, sorry. I am beginner for use git. And after update my fork I to does wrong rebase. Can you please say me like I am can to do correctly it?

**Review by gastaldi — APPROVED** [review]
Works as advertised. Good job!

**Comment by xumk** [comment]
@gmest @geoand approve please, for PR to be counted in hacktoberfest

**Pull request merged**